table/tables/tables.go

```diff
@@ -987,33 +985,59 @@ func (t *TableCommon) AllocHandleIDs(ctx sessionctx.Context, n uint64) (int64, i
 	return base, maxID, nil
 }
 
-// OverflowShardBits checks whether the rowID overflow `1<<(64-shardRowIDBits-1) -1`.
-func OverflowShardBits(rowID int64, shardRowIDBits uint64) bool {
-	mask := (1<<shardRowIDBits - 1) << (64 - shardRowIDBits - 1)
-	return rowID&int64(mask) > 0
+// OverflowShardBits checks whether the recordID overflow `1<<(typeBitsLength-shardRowIDBits-1) -1`.
+func OverflowShardBits(recordID int64, shardRowIDBits uint64, typeBitsLength uint64) bool {
+	mask := (1<<shardRowIDBits - 1) << (typeBitsLength - shardRowIDBits - 1)
+	return recordID&int64(mask) > 0
 }
 
-func (t *TableCommon) calcShard(startTS uint64) int64 {
+// CalcShard calculates the shard prefix by hashing the startTS. Make sure OverflowShardBits is false before calling it.
+func CalcShard(shardRowIDBits uint64, startTS uint64, typeBitsLength uint64) int64 {
 	var buf [8]byte
 	binary.LittleEndian.PutUint64(buf[:], startTS)
 	hashVal := int64(murmur3.Sum32(buf[:]))
-	return (hashVal & (1<<t.meta.ShardRowIDBits - 1)) << (64 - t.meta.ShardRowIDBits - 1)
+	return (hashVal & (1<<shardRowIDBits - 1)) << (typeBitsLength - shardRowIDBits - 1)
 }
 
 // Allocator implements table.Table Allocator interface.
-func (t *TableCommon) Allocator(ctx sessionctx.Context) autoid.Allocator {
+func (t *TableCommon) Allocator(ctx sessionctx.Context, allocType autoid.AllocatorType) autoid.Allocator {
+	allAllocs := t.AllAllocators(ctx)
+	for _, a := range allAllocs {
+		if a.GetType() == allocType {
+			return a
+		}
+	}
+	return nil
+}
+
+// AllAllocators implements table.Table AllAllocators interface.
+func (t *TableCommon) AllAllocators(ctx sessionctx.Context) autoid.Allocators {
```

Comment:
can we make it more pretty here (this func)?

Reply:
Done. PTAL